File: 第6章 哈希表/6.2 哈希函数/HashFunction.java
public class HashFunction {
    public static void main(String[] args) {
        System.out.println(hash("abcde", 100));
    }

    /**
     * @brief  Hash function for calculating:
     *         for all i sum(key[n-i-1] * 32^i) mod tableSize
     * @param  key: key of string type
     * @param  tableSize: size of hash table
     * @retval hash value
     */
    public static int hash(String key, int tableSize) {
        int h = 0;          // hash value
        int n = key.length();
        for(int i = 0; i < n; i++) {
            h = (h << 5) + key.charAt(i);
        }
        return h % tableSize;
    }
}

File: 第6章 哈希表/6.2 哈希函数/HashFunction.py
def hash(key, table_size):
    """
        Hash function for calculating:
        for all i sum(key[n-i-1] * 32^i) mod tableSize
        
        Args:
            key (str): key of string type
            table_size (int): size of hash table

        Returns:
            [int]: hash value
    """
    h = 0       # hash value
    for i in range(len(key)):
        h = (h << 5) + ord(key[i])
    return h % table_size

def main():
    print(hash("abcde", 100))

if __name__ == "__main__":
    main()